WordPress/openverse
# Staging database restore DAG: base restore

## Fixes



This is a step towards #1989 but does not address it completely (steps 12+ from [the implementation plan](https://docs.openverse.org/projects/proposals/search_relevancy_sandbox/20230406-implementation_plan_update_staging_database.html#dag) will be added in follow-up PRs).

## Description



This PR adds the initial restore-from-snapshot steps of the staging database restore DAG. See the linked IP above for the full definition.

I recommend reviewing in the following order:
- `constants.py`
- `utils.py`
- `staging_database_restore.py`
- `staging_database_restore_dag.py`

![Screenshot_2023-05-12_15-47-27](https://github.com/WordPress/openverse/assets/10214785/ddd8c721-6624-4e1e-b6ec-7a6b071e570d)

Since this doesn't have the last required steps for the DAG, we won't enable it once this is merged. I felt that splitting this up into several PRs was a better route to take though, otherwise the full DAG would have been huge!

## Testing Instructions


Tests should pass locally. You can also set the `SKIP_STAGING_DATABASE_RESTORE` Airflow Variable to `true` and kick off the DAG to see that it should skip everything.

> **Warning**
> If you want to test the actual restore operations, you can do so by setting the following environment variables:
> `AWS_RDS_CONN_ID=aws_rds`
> `AIRFLOW_CONN_AWS_RDS=aws://<access-key>:<access-secret>@?region_name=us-east-1`
> 
> This **will** run the restore & delete operations, so be advised!

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
